datalad/plugin/addurls.py

```diff
@@ -0,0 +1,654 @@
+    format_filename = partial(
+        RepFormatter(colidx_to_name, missing_value).format,
+        filename_format)
+    subpaths = _format_filenames(format_filename, rows, infos)
+    return infos, subpaths
+
+
+def dlplugin(dataset=None, url_file=None, input_type="ext",
+             url_format="{0}", filename_format="{1}",
+             exclude_autometa=None, meta=None,
+             message=None, dry_run=False, fast=False,
+             ifexists=None, missing_value=None):
+    """Create and update a dataset from a list of URLs.
+
+    Parameters
+    ----------
+    dataset : Dataset
+        Add the URLs to this dataset (or possibly subdatasets of this
+        dataset).  An empty or non-existent directory is passed to
+        create a new dataset.  New subdatasets can be specified with
+        `filename_format`.
+    url_file : str
+        A file that contains URLs or information that can be used to
+        construct URLs.  Depending on the value of `input_type`, this
+        should be a CSV file (with a header as the first row) or a
+        JSON file (structured as a list of objects with string
+        values).
+    input_type : {"ext", "csv", "json"}, optional
+        Whether `url_file` should be considered a CSV file or a JSON
+        file.  The default value, "ext", means to consider `url_file`
+        as a JSON file if it ends with ".json".  Otherwise, treat it
+        as a CSV file.
+    url_format : str, optional
+        A format string that specifies the URL for each entry.  This
+        value is similar to a normal Python format string where the
+        names from `url_file` (column names for a CSV or properties
+        for JSON) are available as placeholders.  If `url_file` is a
+        CSV file, a positional index can also be used (i.e., "{0}" for
+        the first column).  Note that a placeholder cannot contain a
+        ':' or '!'.
+    filename_format : str, optional
+        Like `url_format`, but this format string specifies the file to
+        which the URL's content will be downloaded.  The file name may
+        contain directories.  The separator "//" can be used to indicate
+        that the left-side directory should be created as a new
+        subdataset.
+
+        In addition to the placeholders described in `url_format`, there
+        are a few special placeholder.
+
+          - "_repindex"
+
+            The constructed file names must be unique across all fields
+            rows.  To avoid collisions, the special placeholder
+            "_repindex" can added to the formatter.  It value will start
+            at 0 and increment every time a file name repeats.
+
+          - "_urlN"
+
+            Each part of the formatted URL is available.  For example,
+            in "http://datalad.org/for/git-users", "_url0" and "_url1"
+            would map to "for" and "git-users", respectively.
+
+            The final part is also available as "_url_basename".
```

Comment:
may be because of entries with missing urls from the target usecase (see other comment) I am getting
```shell
$> datalad plugin addurls url_file=phenotype_mri.csv url_format='{MRIs}' filename_format='{project}/sub-{participant_id}/ses-{session_id}/{_url_basename}' dry_run=True
[ERROR  ] '_url_basename' [string.py:get_value:605] (KeyError)
```
?